arangod/IResearch/IResearchFilterFactory.cpp

```diff
@@ -2186,184 +2302,223 @@ arangodb::Result fromFuncStartsWith(irs::boolean_filter* filter, QueryContext co
 
-    if (arangodb::iresearch::SCOPED_VALUE_TYPE_STRING != patternValue.type()) {
-      auto message = "'LIKE' AQL function: 2nd argument has invalid type '"s + ScopedAqlValue::typeString(patternValue.type()).c_str() + "' (string expected)";
-      LOG_TOPIC("bd9c1", WARN, arangodb::iresearch::TOPIC) << message;
-      return {TRI_ERROR_BAD_PARAMETER, message};
-    }
+    TRI_ASSERT(filterCtx.analyzer);
+    kludge::mangleStringField(name, filterCtx.analyzer);
 
-    if (!patternValue.getString(pattern)) {
-      auto message = "'LIKE' AQL function: Unable to parse 2nd argument as string";
-      LOG_TOPIC("e1ebc", WARN, arangodb::iresearch::TOPIC) << message;
-      return {TRI_ERROR_BAD_PARAMETER, message};
+    auto& wildcardFilter = filter->add<irs::by_wildcard>();
+    wildcardFilter.scored_terms_limit(scoringLimit);
+    wildcardFilter.field(std::move(name));
+    wildcardFilter.boost(filterCtx.boost);
+    wildcardFilter.term(pattern);
+  }
+
+  return {};
+}
+
+// LEVENSHTEIN_MATCH(<attribute>, <target>, <max-distance> [, <include-transpositions>])
+arangodb::Result fromFuncLevenshteinMatch(
+    char const* funcName,
+    irs::boolean_filter* filter,
+    QueryContext const& ctx,
+    FilterContext const& filterCtx,
+    arangodb::aql::AstNode const& args) {
+  TRI_ASSERT(funcName);
+
+  if (!args.isDeterministic()) {
+    return error::nondeterministicArgs(funcName);
+  }
+
+  auto const argc = args.numMembers();
+
+  if (argc < 3 || argc > 4) {
+    return error::invalidArgsCount<error::Range<3, 4>>(funcName);
+  }
+
+  // 1st argument defines a field
+  auto const* field =
+      arangodb::iresearch::checkAttributeAccess(args.getMemberUnchecked(0), *ctx.ref);
+
+  if (!field) {
+    return error::invalidAttribute(funcName, 1);
+  }
+
+  // 2nd argument defines a target
+  ScopedAqlValue targetValue;
+  irs::string_ref target;
+  arangodb::Result res = evaluateArg(target, targetValue, funcName, args, 1, filter, ctx);
+
+  if (!res.ok()) {
+    return res;
+  }
+
+  ScopedAqlValue tmpValue; // can reuse value for int64_t and bool
+
+  // 3rd argument defines a max distance
+  int64_t maxDistance = 0;
+  res = evaluateArg(maxDistance, tmpValue, funcName, args, 2, filter, ctx);
+
+  if (!res.ok()) {
+    return res;
+  }
+
+  // optional 4th argument defines transpositions
+  bool withTranspositions = false;
+
+  if (4 == argc) {
+    res = evaluateArg(withTranspositions, tmpValue, funcName, args, 3, filter, ctx);
+
+    if (!res.ok()) {
+      return res;
     }
   }
 
+  if (!withTranspositions && maxDistance > MAX_LEVENSHTEIN_DISTANCE) {
+    return {
+      TRI_ERROR_BAD_PARAMETER,
+      "'"s.append(funcName).append("' AQL function: max Levenshtein distance must be in range [0, 4]")
+    };
+  } else if (withTranspositions && maxDistance > MAX_DAMERAU_LEVENSHTEIN_DISTANCE) {
+    return {
+      TRI_ERROR_BAD_PARAMETER,
+      "'"s.append(funcName).append("' AQL function: max Damerau-Levenshtein distance must be in range [0, 3]")
```

Comment:
Maybe use std::to_string(MAX_DAMERAU_LEVENSHTEIN_DISTANCE) to denote range bounds ?